File: x402stall/pay.py
"""Simulated x402 challenge / fulfill. HMAC proof, integer amounts."""

from __future__ import annotations

import hashlib
import hmac
import json
import secrets
import time
from dataclasses import dataclass

SIM_SECRET_ENV = "X402_SIM_SECRET"
DEFAULT_AMOUNT = 1000  # integer atomic USDC units (0.001 USDC if 6 decimals)


@dataclass(frozen=True)
class Challenge:
    nonce: str
    amount: int
    pay_to: str
    ts: int

    def to_402_body(self) -> dict:
        return {
            "x402Version": 1,
            "error": "Payment required",
            "accepts": [
                {
                    "scheme": "exact",
                    "network": "sim",
                    "maxAmountRequired": str(self.amount),
                    "payTo": self.pay_to,
                    "asset": "USDC",
                    "extra": {"nonce": self.nonce, "ts": self.ts},
                }
            ],
        }


def make_challenge(amount: int = DEFAULT_AMOUNT, pay_to: str = "sim:stall") -> Challenge:
    if amount <= 0:
        raise ValueError("amount must be positive integer")
    return Challenge(nonce=secrets.token_hex(16), amount=int(amount), pay_to=pay_to, ts=int(time.time()))


def proof_for(secret: str, nonce: str, amount: int) -> str:
    msg = f"{nonce}:{amount}".encode("utf-8")
    return hmac.new(secret.encode("utf-8"), msg, hashlib.sha256).hexdigest()


def verify_payment(secret: str, nonce: str, amount: int, proof: str) -> bool:
    expected = proof_for(secret, nonce, amount)
    return hmac.compare_digest(expected, proof)


def parse_payment_header(raw: str) -> dict:
    data = json.loads(raw)
    if not isinstance(data, dict) or "nonce" not in data or "proof" not in data:
        raise ValueError("payment must be JSON with nonce and proof")
    return data

File: src/pay.ts
import { createHmac, randomBytes, timingSafeEqual } from "node:crypto";

export const SIM_SECRET_ENV = "X402_SIM_SECRET";
export const DEFAULT_AMOUNT = 1000;

export type Challenge = {
  nonce: string;
  amount: number;
  payTo: string;
  ts: number;
};

export function simSecret(): string {
  return process.env[SIM_SECRET_ENV] ?? "dev-only-sim-secret-not-a-wallet";
}

export function makeChallenge(amount: number = DEFAULT_AMOUNT, payTo = "sim:stall"): Challenge {
  if (!Number.isInteger(amount) || amount <= 0) {
    throw new Error("amount must be a positive integer");
  }
  return {
    nonce: randomBytes(16).toString("hex"),
    amount,
    payTo,
    ts: Math.floor(Date.now() / 1000),
  };
}

export function proofFor(secret: string, nonce: string, amount: number): string {
  return createHmac("sha256", secret).update(`${nonce}:${amount}`).digest("hex");
}

export function verifyPayment(secret: string, nonce: string, amount: number, proof: string): boolean {
  const expected = proofFor(secret, nonce, amount);
  const a = Buffer.from(expected, "hex");
  const b = Buffer.from(proof, "hex");
  if (a.length !== b.length) {
    return false;
  }
  // Constant-time compare: a naive `===` leaks via timing if an attacker forges proofs.
  return timingSafeEqual(a, b);
}

export function challengeBody(ch: Challenge): Record<string, unknown> {
  return {
    x402Version: 1,
    error: "Payment required",
    accepts: [
      {
        scheme: "exact",
        network: "sim",
        maxAmountRequired: String(ch.amount),
        payTo: ch.payTo,
        asset: "USDC",
        extra: { nonce: ch.nonce, ts: ch.ts },
      },
    ],
  };
}

export function parsePaymentHeader(raw: string): { nonce: string; proof: string } {
  const data = JSON.parse(raw) as { nonce?: unknown; proof?: unknown };
  if (typeof data.nonce !== "string" || typeof data.proof !== "string") {
    throw new Error("payment must be JSON with nonce and proof");
  }
  return { nonce: data.nonce, proof: data.proof };
}
